Automattic/wp-calypso
# SiteAddressEditor: Validation improvements & Check For Availability Upfront

### Summary

The main objective of this PR is to update the site address editor feature to inform it's users of the availability of their desired site address _before_ the confirmation screen so as to not give any false promise of that availability.
The check for availability will happen whenever there is a break in typing as we assume that this is a reasonable sign that the value in the field is complete.

In order to effectively achieve this, there has been some major work done to the way the existing client-side validation is handled.

### Known issues
- Some of the naming is a bit scattered and off as a result of changing my approach a couple of times while working on this.

### Future Related Work
- Add some placeholder copy in place of the validation message (positive or negative) so as to avoid the sudden jumps in UI between those messages.

### Testing

Validation for this single input form has been a little tricky and there are many caveats to cover in the testing.
Also, you'll need to be working with a patched( D11150 ) and sandboxed WPCOM instance to allow the endpoint to work.

Some overarching things to watch out for:
- The forms submit button should only be submittable when all validation and availability checks are **positive**.
- During server checks, the submit button should show a busy state.
- Any other time (invalid but not checking against the server) the button should be disabled. 
- There are no discrepancies or race conditions between the message we show and the state of the input.

Firstly, lets test cases where the input is too short (`<4`) or too long (`>50`).
  - You'll notice that there is no validation message shown for an empty field both in it's starting position, and after deleting all of the fields content.
  - Start typing anything between 1 and 3 characters.
    - You should see a validation message that suggests that the length of the input is not valid.
    - Notice that the validation message isn't shown immediately but instead there is a slight [debounce](https://lodash.com/docs#debounce)d delay. This is a UX decision to avoid showing an error state as the form user types their first letters. If they stop typing we're assume that their intent is submit the form.
    - The submit button should be disabled during the debounce and when the input is not long enough
    - How does this 'feel', is the timing right or should it be shorter/longer?
  - Type some more characters, taking the length of the input over 4 characters.
    - The `<4` validation message, if already shown, should now be hidden.
  - Backspace some characters to bring the length back down to between 1 and 3.
    - You should see the form behave as before, showing a delayed validation message regarding the input length.
  - Try typing too many characters (`>50`) For your convenience, here's an artisanal address that'll do the trick:`thissiteaddressisagoodexampleofonethatisabittoolong`.
    - Note that the same error message is shown, effectively communicating that the problem is with the length of the address.
  - Backspace a couple of letters to bring the input to less-than 50.
    - note that this specific validation message clears.

Next, take a look at invalid characters.
  - Starting with an input with 4 or more characters to avoid the previous length related validation, type any basic invalid character (<kbd>!</kbd><kbd>@</kbd><kbd>£</kbd><kbd>$</kbd><kbd>%</kbd><kbd>^</kbd><kbd>&</kbd><kbd>*</kbd><kbd>,</kbd> etc.).
    - You should see a validation message warning against this invalid character.
    - Notice that this validation message is shown immediately with no debouncing.
  - Remove this character (and any other invalids) to clear the validation message.
  - Try entering a [diacritic](https://en.wikipedia.org/wiki/Diacritic) character `é` (<kbd>option</kbd>+<kbd>e</kbd>,<kbd>e</kbd>) or `ç` (<kbd>option</kbd>+<kbd>c</kbd>,<kbd>c</kbd>) for example...
    - Note that diacritic characters also cause the same validation message to show.

Next, lets look at server-side validation. As client side validation covers most of the issues that the server would otherwise catch (empty input, input thats too short or long, input with invalid characters) we're left with only one major concern - whether or not the desired site address is available.
  - First, type in an address that is valid, but already taken. `thisaddressistaken` works a treat!
    - Notice that a validation message is shown that communicates that this address is already in use.
    - The behaviour here should be almost exactly the same as the length validation in that it has a delay that debounces as you type so as to not be over eager with the checking of the availability of the address. The only difference being that after the debounce there is a request to the server.
    - During said request, you should notice that the submit button shows a 'busy' state.
    - You should also notice that when the address entered is not available the the submit button has a disabled state - greyed out and nonfunctional.
  - Add some extra random characters until you find an address that is available.
    - You should now see a message that positively validates the availability of this address.
    - You should also see that the submit button now shows it's regular ready-to-use state and should now be actionable (don't worry, there is a confirmation step before the rename actually happens).
